task/control-plane-agent/src/mgs_common.rs

```diff
@@ -33,30 +40,43 @@ impl MgsCommon {
         Ok(DiscoverResponse { sp_port: port })
     }
 
+    pub(crate) fn identity(&self) -> VpdIdentity {
+        if let Some(identity) = *self.identity.borrow() {
+            return identity;
+        }
+
+        let id = identity();
+        let mut cached = self.identity.borrow_mut();
+        *cached = Some(id);
+        id
+    }
+
     pub(crate) fn sp_state(
         &mut self,
         update: &SpUpdate,
         power_state: PowerState,
     ) -> Result<SpState, SpError> {
         ringbuf_entry!(Log::MgsMessage(MgsMessage::SpState));
 
-        // TODO Replace with the real serial number once it's available; for now
-        // use the stm32 96-bit uid
-        let mut serial_number = [0; 16];
-        for (to, from) in serial_number.iter_mut().zip(
-            drv_stm32xx_uid::read_uid()
-                .iter()
-                .flat_map(|x| x.to_be_bytes()),
-        ) {
-            *to = from;
-        }
+        let id = self.identity();
 
-        Ok(SpState {
-            serial_number,
+        let mut state = SpState {
+            serial_number: [0; 32],
+            model: [0; 32],
+            revision: id.revision,
+            hubris_archive_id: kipc::read_image_id().to_le_bytes(),
             version: update.current_version(),
             power_state,
             rot: rot_state(update.sprot_task()),
-        })
+        };
+
+        let n = usize::min(state.serial_number.len(), id.serial.len());
```

Comment:
It would be nice if we could somehow report that we are truncating the serial number.

Reply:
Hmm. How about instead I add a static assertion that we haven't? Both these field sizes are known at compile time.

Reply:
Added these in 72fc2343